drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -543,13 +543,104 @@ static const GLenum gl_primitive[] = {
 
 void RasterizerCanvasGLES2::_canvas_item_render_commands(Item *p_item, Item *current_clip, bool &reclip, RasterizerStorageGLES2::Material *p_material) {
 
+	// We will keep track of what state we have set OpenGL to throughout rendering commands,
+	// in order to minimize OpenGL state changes. This struct keeps a record of state and performs
+	// housekeeping.
+	struct StateCanvasItem {
+		StateCanvasItem() {
+			reset();
+			owner = 0;
+		}
+
+		// clear our internal records of state
+		void reset() {
+			complete = false;
+			tiling = false;
+			force_repeat = false;
+			RID_texture = RID();
+			RID_normal = RID();
+			current_tex = 0;
+			color.a = -1; // just to ensure a difference is found
+			type = -1;
+		}
+
+		bool release(int p_command_type) // return whether command type changed
+		{
+			// noop, no state change
+			if (p_command_type == type)
+				return false;
+
+			// this will always occur on the first command item
+			if (type == -1)
+				return true; // nothing to do
+
+			// put opengl back into a 'default' state when finishing a 'batch' of similar commands
+			switch (type) {
+				// only dealing with speeding up rects so far
+				case Item::Command::TYPE_RECT: {
+					// only considering one case for now
+					if (!owner->use_nvidia_rect_workaround) {
+						set_tiling(false);
+						set_force_repeat(false);
+						glBindBuffer(GL_ARRAY_BUFFER, 0);
+						glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+					}
+				} break;
+			}
+
+			reset();
+
+			return true;
+		}
+
+		void set_force_repeat(bool p_repeat) {
+			// noop
+			if (p_repeat == force_repeat)
+				return;
+
+			owner->state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_FORCE_REPEAT, p_repeat);
+			force_repeat = p_repeat;
+		}
+
+		void set_tiling(bool p_tile) {
+			// noop?
+			if (p_tile == tiling)
+				return;
+
+			if (p_tile) {
+				glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
+				glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
+			} else {
+				glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+				glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+			}
+			tiling = p_tile;
+		}
+
+		int type; // Item::Command::Type
```

Comment:
Should be Item::Command type instead of int

Reply:
Only reason I used an int at the time was because I needed a way or specifying 'unset' (and used -1) as the enum didn't have a NULL value. I'll have a look tomorrow whether it is possible to add to the enum without affecting other bits of the codebase, as agreed that would be better.

Reply:
Ah, may be best to add a "state_dirty" boolean or something similar then. The benefit of using the Enum is that we get the compiler to enforce a lot of static error checking stuff.

Reply:
I had a look at putting a null value into the enum (as this will be the fastest at runtime and clear) and it was doable, so I have gone ahead with this, although it does mean adding extra case statements in a couple of other files to prevent compile warnings (switch not using all the enum values).

It doesn't look like the extra value in the enum will affect anything else (there is no use of the enum for determining array sizes for example), and shouldn't have any significant effect on jump tables. I've put the null value as the first in the enum (as is usually tradition). It could be named something else like 'TYPE_UNSET', but null seems quite language agnostic.